### ISSUE TITLE
并行速度变慢

### ISSUE THREAD

**Issue by eragesrgeha2** [BaoBao0926/ASP-VMUNet#3]
作者你好，我想问一下您有试过并行训练速度更慢甚至显存需求更大情况吗，包括VMunet，我使用并行方式后参数虽然下降但是训练变慢很多

**Comment by BaoBao0926** [comment]
I have not tried to use multi-GPU parallel training on my code. I remember training on a single GPU(RTX4090) on this dataset is fast (maybe 3-4 hours. The smallest model seems to take less than half hour to run on the PH2 dataset.).

If you really want to conduct multi-GPU training, I suggest you look for the source code of Ultralight-VMUNet or VMUNet(this work's baseline) and give it a try to see if it can be trained with multiple GPUs. I remember that when I was initially running the source code of Ultralight-VMUNet, it can conduct multi-GPU training.

By the way, you mentioned that your parameters reduced. This is quite strange. Usually, multi-GPU training is that if a single GPU have a batch size of 20, the total batch size would be 200 if you use 10 GPUs. In this case, the model parameters in each GPU remain the same (and do not decrease). 

Thank you for your attention to my work. If you have any further questions, please feel free to ask.

**Comment by eragesrgeha2** [comment]
你好，我想我可能没有表述清楚。我指的是把输入分为4份再送去vss模块也就是像pvm那样，并非是指并行训练。虽然把输入分成4份，也就是一个vss块是四分之一输入，参数下降了很多。但是训练时长明显变长了，如果把vmunet中的vss全部换成这种并行的vss，我的模型甚至会溢出显存。这在我看来很不合理，我想请问作者有没有遇到过这种情况，谢谢您的解答



---原始邮件---
发件人: "Muyi ***@***.***&gt;
发送时间: 2025年8月21日(周四) 晚上11:00
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [BaoBao0926/ASP-VMUNet] 并行速度变慢 (Issue #3)


BaoBao0926 left a comment (BaoBao0926/ASP-VMUNet#3)
 
I have not tried to use multi-GPU parallel training on my code. I remember training on a single GPU(RTX4090) on this dataset is fast (maybe 3-4 hours. The smallest model seems to take less than half hour to run on the PH2 dataset.).
 
If you really want to conduct multi-GPU training, I suggest you look for the source code of Ultralight-VMUNet or VMUNet(this work's baseline) and give it a try to see if it can be trained with multiple GPUs. I remember that when I was initially running the source code of Ultralight-VMUNet, it can conduct multi-GPU training.
 
By the way, you mentioned that your parameters reduced. This is quite strange. Usually, multi-GPU training is that if a single GPU have a batch size of 20, the total batch size would be 200 if you use 10 GPUs. In this case, the model parameters in each GPU remain the same (and do not decrease).
 
Thank you for your attention to my work. If you have any further questions, please feel free to ask.

**Comment by BaoBao0926** [comment]
I didn't encounter this situation. This looks very strange.

Do you have run the code of Ultralight-VMUNet? Does it work?

**Comment by eragesrgeha2** [comment]
好的，谢谢您&nbsp; 我只是在Vmunet和vmamba测试了一下确实训练变慢了。我再去测试一下找一下原因。感谢您的解答



---原始邮件---
发件人: "Muyi ***@***.***&gt;
发送时间: 2025年8月21日(周四) 晚上11:10
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [BaoBao0926/ASP-VMUNet] 并行速度变慢 (Issue #3)


BaoBao0926 left a comment (BaoBao0926/ASP-VMUNet#3)
 
I didn't encounter this situation. This looks very strange.
 
Do you have run the code of Ultralight-VMUNet? Does it work?

**Comment by chaelf** [comment]
> 指的是把输入分为4份再送去vss模块也就是像pvm那样，并非是指并行训练。虽然把输入分成4份，也就是一个vss块是四分之一输入，参数下降了很多。但是训练时长明显变长了，如果把vmunet中的vss全部换成这种并行的vss，我的模型甚至会溢出显存。这在我看来很不合理，我想请问作者有没有遇到过这种情况，谢谢您的解答

哥们你也是看的Parallel Mamba那篇文章吧，那篇文章就是这样的，通道分组就是会导致参数量下降但是训练时长显著增加

**Comment by eragesrgeha2** [comment]
是的，后面我的卡加其他模块跑不起来就放弃那个创新点了。但是我有点不太理解他们参数和flops降低训练时长确大大增加还有什么意义这种工作



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2025年9月11日(周四) 中午1:59
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [BaoBao0926/ASP-VMUNet] 并行速度变慢 (Issue #3)


chaelf left a comment (BaoBao0926/ASP-VMUNet#3)
  
指的是把输入分为4份再送去vss模块也就是像pvm那样，并非是指并行训练。虽然把输入分成4份，也就是一个vss块是四分之一输入，参数下降了很多。但是训练时长明显变长了，如果把vmunet中的vss全部换成这种并行的vss，我的模型甚至会溢出显存。这在我看来很不合理，我想请问作者有没有遇到过这种情况，谢谢您的解答
  
哥们你也是看的Parallel Mamba那篇文章吧，那篇文章就是这样的，通道分组就是会导致参数量下降但是训练时长显著增加

**Comment by chaelf** [comment]
> 是的，后面我的卡加其他模块跑不起来就放弃那个创新点了。但是我有点不太理解他们参数和flops降低训练时长确大大增加还有什么意义这种工作
> […](#)

他做这个可能就是单纯为了发论文，故事吹的天花乱坠，实际上跑出来的性能连22年的MALUNet都不一定跑得过

**Comment by BaoBao0926** [comment]
我最近想了一下这个问题，我想可能的原因是因为，
我们现在把input(假设维度是 [B, H, W ,C])拆成了4份（现在是四个 [B,H,W,C/4] ），然后一份一份进行运算，调用4次kernel，每一次都是比较小的，这种矩阵运算速度可能就不是很快。代码如下

```
x1, x2, x3, x4 = torch.chunk(input, 4, dim=-1)
x_mamba1 = self.mamba(x1) + self.skip_scale * x1
x_mamba2 = self.mamba(x2) + self.skip_scale * x2
x_mamba3 = self.mamba(x3) + self.skip_scale * x3
x_mamba4 = self.mamba(x4) + self.skip_scale * x4
x_mamba = torch.cat([x_mamba1, x_mamba2, x_mamba3, x_mamba4], dim=-1)
```

有一种可能的办法是，我们把input改成一个大矩阵，维度是 [4, B, H, W, C/4], 这样一个大矩阵可能就可以更加充分利用GPU进行加速，但是似乎需要改Mamba底层的ssm的代码(不太确定)，实现上比较麻烦。或者改成[B*4, H, W, C/4],可能就不需要改太多东西了。

只是一种可能吧，因为理论上并没有多计算量，所以不太可能计算速度下降非常多。

类似的代码如下

```
# input: [B, H, W, C]
B, H, W, C = input.shape

# Step 1: reshape
# [B, H, W, C] -> [B, H, W, 4, C/4] -> [B*4, H, W, C/4]
x = input.reshape(B, H, W, 4, C // 4)
x = x.permute(3, 0, 1, 2, 4).reshape(B * 4, H, W, C // 4)

# Step 2: input this big matrix into Mamba one time 
x_mamba = self.mamba(x) + self.skip_scale * x   # [B*4, H, W, C/4]

# Step 3: reshape to original dim
# [B*4, H, W, C/4] -> [4, B, H, W, C/4] -> [B, H, W, C]
x_mamba = x_mamba.reshape(4, B, H, W, C // 4)
x_mamba = x_mamba.permute(1, 2, 3, 0, 4).reshape(B, H, W, C)
```

**Comment by BaoBao0926** [comment]
> 我最近想了一下这个问题，我想可能的原因是因为， 我们现在把input(假设维度是 [B, H, W ,C])拆成了4份（现在是四个 [B,H,W,C/4] ），然后一份一份进行运算，调用4次kernel，每一次都是比较小的，这种矩阵运算速度可能就不是很快。代码如下
> 
> ```
> x1, x2, x3, x4 = torch.chunk(input, 4, dim=-1)
> x_mamba1 = self.mamba(x1) + self.skip_scale * x1
> x_mamba2 = self.mamba(x2) + self.skip_scale * x2
> x_mamba3 = self.mamba(x3) + self.skip_scale * x3
> x_mamba4 = self.mamba(x4) + self.skip_scale * x4
> x_mamba = torch.cat([x_mamba1, x_mamba2, x_mamba3, x_mamba4], dim=-1)
> ```
> 
> 有一种可能的办法是，我们把input改成一个大矩阵，维度是 [4, B, H, W, C/4], 这样一个大矩阵可能就可以更加充分利用GPU进行加速，但是似乎需要改Mamba底层的ssm的代码(不太确定)，实现上比较麻烦。或者改成[B*4, H, W, C/4],可能就不需要改太多东西了

**Comment by BaoBao0926** [comment]
> > 我最近想了一下这个问题，我想可能的原因是因为， 我们现在把input(假设维度是 [B, H, W ,C])拆成了4份（现在是四个 [B,H,W,C/4] ），然后一份一份进行运算，调用4次kernel，每一次都是比较小的，这种矩阵运算速度可能就不是很快。代码如下
> > ```
> > x1, x2, x3, x4 = torch.chunk(input, 4, dim=-1)
> > x_mamba1 = self.mamba(x1) + self.skip_scale * x1
> > x_mamba2 = self.mamba(x2) + self.skip_scale * x2
> > x_mamba3 = self.mamba(x3) + self.skip_scale * x3
> > x_mamba4 = self.mamba(x4) + self.skip_scale * x4
> > x_mamba = torch.cat([x_mamba1, x_mamba2, x_mamba3, x_mamba4], dim=-1)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 有一种可能的办法是，我们把input改成一个大矩阵，维度是 [4, B, H, W, C/4], 这样一个大矩阵可能就可以更加充分利用GPU进行加速，但是似乎需要改Mamba底层的ssm的代码(不太确定)，实现上比较麻烦。或者改成[B*4, H, W, C/4],可能就不需要改太多东西了

**Comment by BaoBao0926** [comment]
> > > 我最近想了一下这个问题，我想可能的原因是因为， 我们现在把input(假设维度是 [B, H, W ,C])拆成了4份（现在是四个 [B,H,W,C/4] ），然后一份一份进行运算，调用4次kernel，每一次都是比较小的，这种矩阵运算速度可能就不是很快。代码如下
> > > ```
> > > x1, x2, x3, x4 = torch.chunk(input, 4, dim=-1)
> > > x_mamba1 = self.mamba(x1) + self.skip_scale * x1
> > > x_mamba2 = self.mamba(x2) + self.skip_scale * x2
> > > x_mamba3 = self.mamba(x3) + self.skip_scale * x3
> > > x_mamba4 = self.mamba(x4) + self.skip_scale * x4
> > > x_mamba = torch.cat([x_mamba1, x_mamba2, x_mamba3, x_mamba4], dim=-1)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 有一种可能的办法是，我们把input改成一个大矩阵，维度是 [4, B, H, W, C/4], 这样一个大矩阵可能就可以更加充分利用GPU进行加速，但是似乎需要改Mamba底层的ssm的代码(不太确定)，实现上比较麻烦。或者改成[B*4, H, W, C/4],可能就不需要改太多东西了。

只是一种可能，因为理论上并没有多额外的计算量，不应该明显的计算速度下降

**Comment by eragesrgeha2** [comment]
好的，谢谢您的解答



---原始邮件---
发件人: "Muyi ***@***.***&gt;
发送时间: 2025年9月17日(周三) 凌晨1:48
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [BaoBao0926/ASP-VMUNet] 并行速度变慢 (Issue #3)


BaoBao0926 left a comment (BaoBao0926/ASP-VMUNet#3)
    
我最近想了一下这个问题，我想可能的原因是因为， 我们现在把input(假设维度是 [B, H, W ,C])拆成了4份（现在是四个 [B,H,W,C/4] ），然后一份一份进行运算，调用4次kernel，每一次都是比较小的，这种矩阵运算速度可能就不是很快。代码如下
 x1, x2, x3, x4 = torch.chunk(input, 4, dim=-1) x_mamba1 = self.mamba(x1) + self.skip_scale * x1 x_mamba2 = self.mamba(x2) + self.skip_scale * x2 x_mamba3 = self.mamba(x3) + self.skip_scale * x3 x_mamba4 = self.mamba(x4) + self.skip_scale * x4 x_mamba = torch.cat([x_mamba1, x_mamba2, x_mamba3, x_mamba4], dim=-1)  
有一种可能的办法是，我们把input改成一个大矩阵，维度是 [4, B, H, W, C/4], 这样一个大矩阵可能就可以更加充分利用GPU进行加速，但是似乎需要改Mamba底层的ssm的代码(不太确定)，实现上比较麻烦。或者改成[B*4, H, W, C/4],可能就不需要改太多东西了。
    
只是一种可能，因为理论上并没有多额外的计算量，不应该明显的计算速度下降

**Comment by eragesrgeha2** [comment]
你好，教授&nbsp; 我看您网络中使用了Se模块，我想请教一下我的卷积分支后使用的是silu，我想接一个Se模块，但是我的Se模块内部的FC层用什么激活函数呢是否需要跟他前面的卷积一样用silu，但是我看论文Se好像没有用silu的不知道可不可以，还是说用relu，这样一个分支中既有silu又有relu不知道是否合适



---原始邮件---
发件人: "Muyi ***@***.***&gt;
发送时间: 2025年9月17日(周三) 凌晨1:48
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [BaoBao0926/ASP-VMUNet] 并行速度变慢 (Issue #3)


BaoBao0926 left a comment (BaoBao0926/ASP-VMUNet#3)
    
我最近想了一下这个问题，我想可能的原因是因为， 我们现在把input(假设维度是 [B, H, W ,C])拆成了4份（现在是四个 [B,H,W,C/4] ），然后一份一份进行运算，调用4次kernel，每一次都是比较小的，这种矩阵运算速度可能就不是很快。代码如下
 x1, x2, x3, x4 = torch.chunk(input, 4, dim=-1) x_mamba1 = self.mamba(x1) + self.skip_scale * x1 x_mamba2 = self.mamba(x2) + self.skip_scale * x2 x_mamba3 = self.mamba(x3) + self.skip_scale * x3 x_mamba4 = self.mamba(x4) + self.skip_scale * x4 x_mamba = torch.cat([x_mamba1, x_mamba2, x_mamba3, x_mamba4], dim=-1)  
有一种可能的办法是，我们把input改成一个大矩阵，维度是 [4, B, H, W, C/4], 这样一个大矩阵可能就可以更加充分利用GPU进行加速，但是似乎需要改Mamba底层的ssm的代码(不太确定)，实现上比较麻烦。或者改成[B*4, H, W, C/4],可能就不需要改太多东西了。
    
只是一种可能，因为理论上并没有多额外的计算量，不应该明显的计算速度下降

**Comment by BaoBao0926** [comment]
我感觉用silu还是relu应该是都是可以的，无所谓。

一个分支里面有不同的激活函数也没有太大的问题

如果这不是你的核心，我感觉就follow原文用relu就行了（当然用silu也没有问题），我感觉效果大差不大。如果这是你纠结的点/你认为重要的点/你想探究的点，那你做一下消融实验，看看哪一个激活函数效果好就行了

By the way, 我现在是CMU的first-year master，说教授可太抬举我了😄

**Comment by eragesrgeha2** [comment]
好的，谢谢



---原始邮件---
发件人: "Muyi ***@***.***&gt;
发送时间: 2025年11月2日(周日) 凌晨3:45
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [BaoBao0926/ASP-VMUNet] 并行速度变慢 (Issue #3)


BaoBao0926 left a comment (BaoBao0926/ASP-VMUNet#3)
 
我感觉用silu还是relu应该是都是可以的，无所谓。
 
一个分支里面有不同的激活函数也没有太大的问题
 
如果这不是你的核心，我感觉就follow原文用relu就行了（当然用silu也没有问题），我感觉效果大差不大。如果这是你纠结的点/你认为重要的点/你想探究的点，那你做一下消融实验，看看哪一个激活函数效果好就行了
 
By the way, 我现在是CMU的first-year master，说教授可太抬举我了😄

**Comment by eragesrgeha2** [comment]
您好教授。我想问一下对于vmunet和您的工作，使用的因果卷积和mamba是1.0.0和1.1.1的，我可以使用更高版本的mamba比如torch2.1+mamba1.1几+conv1d1.1几这样的环境去运行和创造吗，因为我看似乎vmamba使用的是更高版本的但是安装方式不同，网上资料很少。&nbsp; &nbsp;谢谢



---原始邮件---
发件人: "Muyi ***@***.***&gt;
发送时间: 2025年8月21日(周四) 晚上11:10
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [BaoBao0926/ASP-VMUNet] 并行速度变慢 (Issue #3)


BaoBao0926 left a comment (BaoBao0926/ASP-VMUNet#3)
 
I didn't encounter this situation. This looks very strange.
 
Do you have run the code of Ultralight-VMUNet? Does it work?

**Comment by BaoBao0926** [comment]
I think both are ok, as long as you can run it.
